Support for RX IRQ

andreamerello opened the pull request:
Add support for RX IRQ. This is useful because of two reasons:

-  Avoid most useless I2C transactions (due to blind RX polling).
-  Shortened (and hopefully independent by USB-C stack thread polling rate) delay to respond to messages. Note that  the source (at least mine) sends a hardware resets messages and it removes the power if it doesn't receive any request messages after ~25ms since any acknowledged capability advertisement, so extra delay is harmful here..

BTW  The USB-C stack still continues to periodically polling for pending messages, but with this patch the handler is often able to return -ENODATA without waste an I2C transaction (I wonder whether there is some method to tell the USB-C stack not to poll at all since we support interrupts..).

Probably we may want to make the IRQ support *optional* while this patch doesn't care about that, but I wanted to share this early.

This PR indeed contains also another small commit that fixed a compile error due to missing function (logging-related).

andreamerello committed:
fix compile failure due to missing function

andreamerello committed:
add support for RX irq

Signed-off-by: Andrea Merello <andrea.merello@iit.it>

ottojo reviewed (approved):
Thanks for the contribution! IRQ support would have been one of the next TODO items while i worked on this :smile: 

The `pd_packet_type_to_str` function seems to have gotten lost in the "cleanup" commit (haha), here it is:
```c
static const char *pd_packet_type_to_str(enum pd_packet_type t)
{
	switch (t) {
	case PD_PACKET_SOP:
		return "PD_PACKET_SOP";
	case PD_PACKET_SOP_PRIME:
		return "PD_PACKET_SOP_PRIME";
	case PD_PACKET_PRIME_PRIME:
		return "PD_PACKET_PRIME_PRIME";
	case PD_PACKET_DEBUG_PRIME:
		return "PD_PACKET_DEBUG_PRIME";
	case PD_PACKET_DEBUG_PRIME_PRIME:
		return "PD_PACKET_DEBUG_PRIME_PRIME";
	case PD_PACKET_TX_HARD_RESET:
		return "PD_PACKET_TX_HARD_RESET";
	case PD_PACKET_CABLE_RESET:
		return "PD_PACKET_CABLE_RESET";
	case PD_PACKET_TX_BIST_MODE_2:
		return "PD_PACKET_TX_BIST_MODE_2";
	case PD_PACKET_MSG_INVALID:
		return "PD_PACKET_MSG_INVALID";
	}
	return "<invalid pd_packet_type>";
}
```
feel free to add that back, it was useful for me during debugging at least...

I don't have the capacity to test this on hardware (or at all) right now, but i looked over the code and it seems sensible to me.

I agree that making IRQ support optional would be nice, but i would also merge this if it made IRQ mandatory. In retrospect, a bunch of issues i had previously likely came down to sending responses too late...
Let me know if you are adding that anyways, otherwise i will just merge this :+1:

andreamerello committed:
make IRQ optional

Signed-off-by: Andrea Merello <andrea.merello@iit.it>

andreamerello commented:
I've added one extra commit to make IRQ optionals, but I could test it only with IRQ enabled, because on my board it seems that otherwise things are too slow, and the PD protocol timeouts (that is the actual reason I've implemented IRQ support indeed).

ottojo reviewed (changes requested):
looks good! one last request: could you add the new `int-gpios` devicetree property to the example in the readme? i think using the chip with interrupts would be the preferred mode, and the example should show that. thanks!
